kafka-connect-s3/src/main/java/io/confluent/connect/s3/util/S3ErrorUtils.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.connect.s3.util;
+
+import com.amazonaws.AmazonClientException;
+import com.amazonaws.AmazonWebServiceRequest;
+import com.amazonaws.retry.PredefinedRetryPolicies;
+import io.confluent.connect.storage.common.util.StringUtils;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.errors.RetriableException;
+
+import java.io.IOException;
+
+/**
+ * Utilities related to basic S3 error/exception analysis.
+ */
+public class S3ErrorUtils {
+
+  /**
+   * Return whether the given exception is a "retryable" exception.
+   * @param exception The exception to analyze
+   * @return true if the exception is retryable
+   */
+  private static boolean isRetriableException(Throwable exception) {
+    if (exception == null) {
+      return false;
+    }
+
+    // IOException, in many places, is passed the AWS exception
+    // when it is thrown. We search here to check that exception
+    // for the IOException case.  Otherwise, the IOException
+    // is considered not retriable.
+    // Exception: An IOException embedded within an `AmazonClientException`
+    // should be passed via the `AmazonClientException` object
+    // as its parent (as the SDK does), in which case, shouldRetry()
+    // will often find it retriable.
+    for (Throwable cause : ExceptionUtils.getThrowableList(exception)) {
```

Comment:
@lion-simba I am not following why you replaced the recursion here with a loop.  The logic seems different now.  Can you explain?

Reply:
I don't think it is correct to dig more than one level deep into an AWS exception for an IOException.  The SDK does not do this: https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedRetryPolicies.java#L192

Reply:
nevermind, this is ok

Reply:
I just rewrote the original recursive version in a loop form with better "cyclic reference" handling (as suggested above by @kpatelatwork). The logic should have remained exactly the same. Please let me know if it hasn't.

However, @kpatelatwork is suggesting to change original logic so that not only IOException is inspected for AmazonClientException cause (see [above](https://github.com/confluentinc/kafka-connect-storage-cloud/pull/461#discussion_r775041556)). Please discuss with him and let me know your decision.

Reply:
We discussed and I think Chris has raised a PR against your branch. 

It's holidays here now @lion-simba so we will discuss it in the new year.

Reply:
Oh, I didn't notice the PR. Reviewed now - @cjolivier01, please take a look.

Yeah, it's holidays here too. :) Wish you all happy New Year!